core/uploader.go

```diff
@@ -134,22 +110,87 @@ func Upload(input io.Reader, outputURI *url.URL, waitBetweenWrites, writeTimeout
 	}
 
 	// We have to do this final write, otherwise there might be final data that's arrived since the last periodic write
-	if _, err := session.SaveData(context.Background(), "", bytes.NewReader(fileContents), fields, writeTimeout); err != nil {
+	if _, _, err := uploadFileWithBackup(outputURI, fileContents, fields, writeTimeout, false, storageBackupURLs); err != nil {
 		// Don't ignore this error, since there won't be any further attempts to write
 		return nil, fmt.Errorf("failed to write final save: %w", err)
 	}
 	glog.Infof("Completed writing %s to storage", outputURI.Redacted())
 	return nil, nil
 }
 
-func extractThumb(session drivers.OSSession, filename string, segment []byte) error {
+func uploadFileWithBackup(outputURI *url.URL, fileContents []byte, fields *drivers.FileProperties, writeTimeout time.Duration, withRetries bool, storageBackupURLs StorageBackupURLs) (out *drivers.SaveDataOutput, bytesWritten int64, err error) {
+	out, bytesWritten, primaryErr := uploadFile(outputURI, fileContents, fields, writeTimeout, withRetries)
+	if primaryErr == nil {
+		return out, bytesWritten, nil
+	}
+
+	backupURI, err := buildBackupURI(outputURI, storageBackupURLs)
+	if err != nil {
+		glog.Errorf("failed to build backup URL: %v", err)
+		return nil, 0, primaryErr
+	}
+
+	glog.Warningf("Primary upload failed, uploading to backupURL=%s primaryErr=%q", backupURI.Redacted(), primaryErr)
+	return uploadFile(backupURI, fileContents, fields, writeTimeout, withRetries)
+}
+
+func buildBackupURI(outputURI *url.URL, storageBackupURLs StorageBackupURLs) (*url.URL, error) {
+	outputURIStr := outputURI.String()
+	for _, strj := range storageBackupURLs {
+		if strings.HasPrefix(outputURIStr, strj.Primary) {
```

Comment:
Is it even possible that the `outputURIStr` does not contain `strj.Primary`? 🤔 

If not and it's just to double-check, then I'd print an error every time it happens.

Reply:
Yes, the idea is to support any number of storages here. We could have for example different regions recording to different buckets but both `catalyst-uploader` and `catalyst-api` know what is the corresponding backup URL.

The prefix-checking is a core piece of the logic here since the whole backup URL depends on replacing the provided `primary` prefix with the corresponding `backup` string. I don't think it makes sense to attempt a "blind" replace here without checking if the prefix ("base directory") matches the target URL. I also don't think we need to be strict here and only allow uploading to storages that have a backup storage configured. It's fine if there isn't, we just don't retry on the backup.

Notice that we already print an error whenever there's no backup storage for a URL: error returned on L145 and logged on :129.

Let me know if you have any other change suggestions!